### ISSUE TITLE
Why add updates to local commitment tx only after `revoke_and_ack`?

### ISSUE THREAD

**Issue by OrfeasLitos** [lightning/bolts#569]
In BOLT 2 we read: 
> [A] node only adds [...] updates to its own commitment transaction when the remote node acknowledges it has applied them via `revoke_and_ack`.

Why not apply the "add" updates to one's own commitment tx right after sending the `update_add`? This would bring the communication cost that was pointed out in #472 down to 2 rounds again. (I recognize that this cannot be done for "fulfill" updates.)

Is there a security reason not to? Or this is a breaking change and can't be applied for historical reasons?

**Comment by ysangkok** [comment]
An update has not been fully applied before it has been "irrevocably committed" (signature received and previous ctx'es revoked).

Why would you sign anything for your counterparty before they have committed to it? Also, the protocol has to function if the connection drops after the add. The current behaviour is to forget about all changed that aren't acked. What would your suggested alternate protocol do? It becomes a lot more complicated if changed aren't locked in on one side first. It wouldn't be nice to have a protocol where, if HTLCs are sent concurrently in both directions, you'd have to validate the commitment_signed against different transactions, to see how much your counterparty received before they sent this commitment_signed!

It is indeed a breaking change, but I don't think your proposed solution works anyway. So I wouldn't call it historical cruft.

**Comment by OrfeasLitos** [comment]
> Why would you sign anything for your counterparty before they have committed to it?

Applying the update to one's own ctx doesn't produce any signature for the counterparty's ctx.

> The current behaviour is to forget about all changed that aren't acked.

Applying an update to one's own ctx doesn't make one forget their own latest irrevocably commited ctx. In case a connection drops, one can just reset their own ctx to their own latest irrevocably commited ctx.

**Comment by OrfeasLitos** [comment]
As @ysangkok does not currently work on Lightning, can somebody else respond to my original question? I would appreciate a more detailed explanation.

> Why not apply the "add" updates to one's own commitment tx right after sending the `update_add`?

**Comment by t-bast** [comment]
Sorry for the (very) late reply, you probably have a better understanding of the protocol now but here is a brief explanation. I think you're just confused about what "apply" means here, because it's mostly a hint to guide implementations.

What it means is that once you've received your peer's revocation, you can update the commitment txs by applying all the pending updates and discard the previous versions of the commitment (you only need to remember pending htlc details in case your peer publishes a revoked commitment). You can only do that once you've received signatures from your peer (to actually be able to broadcast this version of the commitment) and its revocation (to discard the previous tx).

**Comment by t-bast** [comment]
Honestly I don't understand what your goal is with this issue...this is not a spec _requirement_, your implementation can do whatever it pleases, it just highlights what makes most sense from an implementation's point of view (since it's a not-so-obvious point until you actually implement it), so I don't see why we'd need to change anything in the spec?

Are you implementing this? If you do, feel free to do whatever you'd like if it makes more sense in your model, the spec doesn't prevent that.

**Comment by OrfeasLitos** [comment]
My goal was to understand exactly the security requirements of Lightning. From your last response, I deduce that the spec is here indeed more stringent than needed.

The phrasing "[A] node only adds [...] updates to its own commitment transaction when the remote node acknowledges it has applied them via `revoke_and_ack`" does sound like a requirement.
